app/pb/search.proto

```diff
@@ -0,0 +1,112 @@
+syntax = "proto3";
+
+package org.couchers.api.search;
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/wrappers.proto";
+
+import "pb/api.proto";
+import "pb/communities.proto";
+import "pb/discussions.proto";
+import "pb/groups.proto";
+import "pb/pages.proto";
+
+service Search {
+  // This is a secure service: a user needs to be authenticated and not jailed to call functions here, refer to auth.proto and jail.proto
+
+  rpc Search(SearchReq) returns (SearchRes) {
+    // Search all content
+  }
+
+  rpc UserSearch(UserSearchReq) returns (UserSearchRes) {
+    // Search for users in particular with extra filters, e.g. hosts
+  }
+}
+
+message Area {
+  double lat = 1;
+  double lng = 2;
+  // truncated to between 5000 (5 km) and 200000 (200 km) for privacy
+  double radius = 3; // m
+}
+
+message SearchReq {
+  string query = 1;
+
+  bool include_users = 4;
+  bool include_communities = 5;
+  bool include_groups = 6;
+  bool include_places = 7;
+  bool include_guides = 8;
+  bool include_discussions = 9;
+  // bool include_events = 10;
+
+  // at most one of the following
+  Area search_in_area = 11;
+  int64 search_in_community = 12;
+
+  bool exclude_title = 13;
+  bool exclude_content = 14;
+
+  uint32 page_size = 2;
+  string page_token = 3;
+}
+
+message Result {
+  // score between 0 and 1
+  double relevance = 1;
+
+  oneof result {
+    org.couchers.api.core.User user = 2;
+    org.couchers.api.communities.Community community = 3;
+    org.couchers.api.groups.Group group = 4;
+    org.couchers.api.pages.Page place = 5;
+    org.couchers.api.pages.Page guide = 6;
+    org.couchers.api.discussions.Discussion discussion = 7;
+    // org.couchers.api.events.Event event = 8;
+  }
```

Comment:
The generated type for this could be a bit clunky, as we'd have to check for each result item what kind of result it is to display accordingly on the frontend. I wonder if there's a good way to unify this into some common interface so we can loop through the results easily and display it through some "high level" views? And if the user want more details, maybe it can expand to one of these more detailed entities?

Reply:
I think it's fine, just have `<SearchResultCard>` that is a big switch statement returning `<UserSearchResultCard>`, `<GroupSearchResultCard>`, etc. Or would that have other problems?

Reply:
That's fine, but I was thinking if there's a way to keep the display logic simpler on the frontend then that's always better. May not be possible if each card is going to be quite different though looking at the designs so far... 🤔

Reply:
I see. Yes I think they're intrinsically too different to try and unify.

Reply:
Yeah, I could return them as repeated users, communities, etc. But you'd have to order them in the frontend then

Reply:
Would it be good, for each search result, to return an URL and a text snippet containing the searched term? To ease the presentation of the search results.

Reply:
> return an URL

URL to where? If you mean to the resultant object (e.g. user), I think that should go inside the `User` object itself instead. I've some thoughts on that.

> a text snippet containing the searched term

You mean like a snippet of the match? I think that'd be a good idea. I was just thinking that it might be annoying to implement :P

Reply:
An URL to the resource according to our URL plan :) Yes a snippet with the search term highlighted like a google search result (client can highlight)

Reply:
> An URL to the resource according to our URL plan :)

Yeah... That should probably be included in the objects themselves. And I've been starting to think we should refactor the whole API to be more resource-oriented.

The only thing is that including a URL is very web-centric. What does it mean for mobile apps or third party apps? Of course we could have our own URI scheme of some form, `couchers://community/1/world` (or just return the path only)?

> Yes a snippet with the search term highlighted like a google search result (client can highlight)

Yeah, I'll add that. Would it be better if the highlighting was through markdown? Or just choose some symbol, `*` or `|`, and highlight sections with that? We can probably use the postgres full-text search to get pretty decent search for now: it doesn't match verbatim, it matches based on word roots or whatever.

Reply:
> That should probably be included in the objects themselves.

But then we lose the whole point that it is hard to deal with oneof in the frontends...

> The only thing is that including a URL is very web-centric.

It's not. Just because an RFC1738 uniform resource locator happens to start with https, doesn't mean any app can parse it and pick out the ID numbers.

Reply:
> return an URL

We won't use this in the frontend. We will have to handle each type of result differently no matter what, and we have route functions now like `routeToUser(userId)`.

> and a text snippet containing the searched term

I wouldn't spend too much time on that, while it would be good for guides/discussions/etc. it isn't in the current design (and we have lots of other stuff to do 😱 )

Reply:
Oh my... I love postgres:

https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-HEADLINE